### PR TITLE
[WebGPU] getBindGroupLayout() is actually supposed to have "create" semantics

### PR DESCRIPTION
**Pull request by litherum**
#### 15c8024a15bf3339de61e0100b9d1317f108d30e
<pre>
[WebGPU] getBindGroupLayout() is actually supposed to have &quot;create&quot; semantics
<a href="https://bugs.webkit.org/show_bug.cgi?id=251732">https://bugs.webkit.org/show_bug.cgi?id=251732</a>
rdar://105030025

Reviewed by Dean Jackson.

This is a partial revert of 259609@main. You would think that, because the name of &quot;getBindGroupLayout()&quot;
starts with the word &quot;get,&quot; that it would have &quot;get&quot; semantics. However, the spec actually explicitly
describes that it has &quot;create&quot; semantics. This patch updates the implementation to have those semantics.

The spec specifically says &quot;A new GPUBindGroupLayout wrapper is returned each time.&quot;

<a href="https://github.com/gpuweb/gpuweb/pull/3804">https://github.com/gpuweb/gpuweb/pull/3804</a> is a PR to the spec to rename the functions, to be more clear
about their behavior.

* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUComputePipelineImpl.cpp:
(PAL::WebGPU::ComputePipelineImpl::getBindGroupLayout):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUComputePipelineImpl.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPURenderPipelineImpl.cpp:
(PAL::WebGPU::RenderPipelineImpl::getBindGroupLayout):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPURenderPipelineImpl.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPUComputePipeline.h:
* Source/WebCore/PAL/pal/graphics/WebGPU/WebGPURenderPipeline.h:
* Source/WebGPU/WebGPU/APIConversions.h:
(WebGPU::releaseToAPI):
* Source/WebGPU/WebGPU/ComputePipeline.h:
* Source/WebGPU/WebGPU/ComputePipeline.mm:
(WebGPU::ComputePipeline::getBindGroupLayout):
(wgpuComputePipelineGetBindGroupLayout):
* Source/WebGPU/WebGPU/RenderPipeline.h:
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::RenderPipeline::getBindGroupLayout):
(wgpuRenderPipelineGetBindGroupLayout):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteComputePipeline.cpp:
(WebKit::RemoteComputePipeline::getBindGroupLayout):
* Source/WebKit/GPUProcess/graphics/WebGPU/RemoteRenderPipeline.cpp:
(WebKit::RemoteRenderPipeline::getBindGroupLayout):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteComputePipelineProxy.cpp:
(WebKit::WebGPU::RemoteComputePipelineProxy::getBindGroupLayout):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteComputePipelineProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteRenderPipelineProxy.cpp:
(WebKit::WebGPU::RemoteRenderPipelineProxy::getBindGroupLayout):
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteRenderPipelineProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/WebGPU/RemoteShaderModuleProxy.cpp:
(WebKit::WebGPU::RemoteShaderModuleProxy::compilationInfo):

Canonical link: <a href="https://commits.webkit.org/259866@main">https://commits.webkit.org/259866@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/44c34279f6c5a5010c35e65874c118436630e359

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106168 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15219 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39001 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115350 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/175426 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110072 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16660 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6404 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98371 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115036 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111926 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12675 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95646 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40211 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94548 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27289 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/81896 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8465 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28641 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8965 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5197 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14581 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48187 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6819 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10501 "Built successfully") | | | 
| | | | |

### CODE REVIEW

File: Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUComputePipelineImpl.cpp [hunk 1]
```diff
@@ -47,9 +47,8 @@ ComputePipelineImpl::~ComputePipelineImpl()
 
 Ref<BindGroupLayout> ComputePipelineImpl::getBindGroupLayout(uint32_t index)
 {
-    return m_bindGroupLayouts.ensure(index, [this, index] {
-        return BindGroupLayoutImpl::create(wgpuComputePipelineGetBindGroupLayout(m_backing, index), m_convertToBackingContext);
-    }).iterator->value;
+    // "A new GPUBindGroupLayout wrapper is returned each time"
```
Comment (mwyrzykowski) [inline Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUComputePipelineImpl.cpp:50]: Is this change really necessary when the only writable member in GPUBindGroupLayout is label?

Can we return the same Ref<BindGroupLayout> inside a new GPUBindGroupLayout to satisfy the specification? Or is that not sufficient as the label will be different and one would expect the label of a GPUBindGroupLayout to go through to the entire device timeline

Comment (litherum) [reply Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUComputePipelineImpl.cpp:50]: I think it's up to us. It's a tradeoff: on one hand, fewer objects / allocations; on the other hand, labels will behave in a sensible manner.

I think we should start with the sensible labels first, and view this as a potential optimization opportunity when we get to the point of trying to optimize things.

**Review by mwyrzykowski — APPROVED** [review]

**Review by grorg — APPROVED** [review]
